src/server/debugcmd.cc

```diff
@@ -68,32 +69,97 @@ struct ValueCompressInfo {
+    string_view type, std::string key, size_t val_size, bool random_value, uint32_t elements,
+    const CommandRegistry& registry, absl::InsecureBitGen* gen) {
+  absl::InlinedVector<string, 5> args;
+  args.push_back(std::move(key));
+
+  const CommandId* cid = nullptr;
+  if (type == "STRING") {
+    cid = registry.Find("SET");
+    args.push_back(GenerateValue(val_size, random_value, gen));
+  } else if (type == "LIST") {
+    cid = registry.Find("LPUSH");
+    for (uint32_t i = 0; i < elements; ++i) {
+      args.push_back(GenerateValue(val_size, random_value, gen));
+    }
+  } else if (type == "SET") {
+    cid = registry.Find("SADD");
+    for (size_t i = 0; i < elements; ++i) {
+      args.push_back(GenerateValue(val_size, random_value, gen));
+    }
+  } else if (type == "HSET") {
+    cid = registry.Find("HSET");
+    for (size_t i = 0; i < elements; ++i) {
+      args.push_back(GenerateValue(val_size / 2, random_value, gen));
+      args.push_back(GenerateValue(val_size / 2, random_value, gen));
+    }
+  } else if (type == "ZSET") {
+    cid = registry.Find("ZADD");
+    for (size_t i = 0; i < elements; ++i) {
+      args.push_back(absl::StrCat((*gen)() % val_size));
+      args.push_back(GenerateValue(val_size, random_value, gen));
+    }
+  } else if (type == "JSON") {
+    cid = registry.Find("JSON.SET");
+    args.push_back("$");
+
+    string json = "{";
+    for (size_t i = 0; i < elements; ++i) {
+      absl::StrAppend(&json, "\"", i, "\":\"", GenerateValue(val_size, random_value, gen), "\",");
+    }
+    json[json.size() - 1] = '}';  // Replace last ',' with '}'
+    args.push_back(json);
+  }
+  return {cid, args};
+}
+
+void DoPopulateBatch(string_view type, string_view prefix, size_t val_size, bool random_value,
+                     int32_t elements, const PopulateBatch& batch, ServerFamily* sf,
+                     ConnectionContext* cntx) {
+  boost::intrusive_ptr<Transaction> local_tx =
+      new Transaction{sf->service().mutable_registry()->Find("EXEC")};
+  local_tx->StartMultiNonAtomic();
+  absl::InlinedVector<MutableSlice, 5> args_view;
+  facade::CapturingReplyBuilder crb;
+  ConnectionContext local_cntx{cntx, local_tx.get(), &crb};
 
   absl::InsecureBitGen gen;
   for (unsigned i = 0; i < batch.sz; ++i) {
     string key = absl::StrCat(prefix, ":", batch.index[i]);
-    string val;
 
-    if (random_value_str) {
-      val = GetRandomHex(gen, val_size);
-    } else {
-      val = absl::StrCat("value:", batch.index[i]);
-      if (val.size() < val_size) {
-        val.resize(val_size, 'x');
-      }
+    auto [cid, args] = GeneratePopulateCommand(type, std::move(key), val_size, random_value,
+                                               elements, *sf->service().mutable_registry(), &gen);
+    if (!cid) {
+      LOG_EVERY_N(WARNING, 10'000) << "Unable to find command, was it renamed?";
+      break;
     }
 
-    auto res = sg.Set(params, key, val);
```

Comment:
So debug populate wasn't transactional before? Now it is

You can save some calculations by making your local transaction a stub transaction, if you are sure that
* You don't want to run transactional logic
* You're on the correct shard

Reply:
that's a good idea, as I also discovered that the current (what you reviewed) implementation doesn't work with a replica connected (it hangs forever as it can't run inline), but a stub fixes that while also updating replica :)
can you kindly review that my changes to use a stub are correct though?

Reply:
Does it update the replica as well? 😄 I assume you need a `ReportWritesSquashedMulti` at the end to send flushing `EXEC` opcodes (or we'll have the same problem as single shard eval). Otherwise the usage looks correct

Reply:
It works because it's not really a MULTI, so no EXEC is required...

Reply:
Ah, yes, I forgor